### ISSUE TITLE
Separate boolean and symbolic relationals

### ISSUE THREAD

**Issue by rlamy** [sympy/sympy#4986]
Well, it seems obvious that a Relational is a Boolean. The problem is that
there are a few places where Expr methods are called on them.

Original issue for #4986: http://code.google.com/p/sympy/issues/detail?id=1887
Original author: https://code.google.com/u/101272611947379421629/
Original owner: https://code.google.com/u/101272611947379421629/

**Comment by micklat** [comment]
I felt a personal need for an Eqn class so I wrote one and found it very useful. I'm new to sympy and not keen on delving deep into this so maybe I'm missing a better way to do this. I'm afraid I'm not very interested in working this into a pull request, but if somebody else would like to pick it up and add it so Sympy, be my guest.

```
import sympy
from sympy import S
import operator

class Eqn(object):
  """
  This class represents an equation. It is meant for reasoning with equations: adding
  equations, subtracting equations, multiplying them, making substitutions etc.

  >>> from sympy import var
  >>> var('x')
  x
  >>> eq1 = Eqn(2*x, 5)
  >>> eq1 += 4
  >>> eq1.solve()
  {x: 5/2}
  """

  __slots__ = ['sides']

  def __init__(self, *sides):
    self.sides = map(S, sides)
  
  def map(self, *fs):
    res = self
    for f in fs:
      res = Eqn(*map(f, res.sides))
    return res

  def __iadd__(self, other):
    self.sides = (self+other).sides
    return self

  def __imul__(self, other):
    self.sides = (self*other).sides
    return self

  def __isub__(self, other):
    self.sides = (self-other).sides
    return self

  def __itruediv__(self, other):
    self.sides = operator.truediv(self,other).sides
    return self
  
  def __rmul__(self, other):
    return self.binary_op(operator.mul, other, self)

  def __radd__(self, other):
    return self.binary_op(operator.add, other, self)
    
  def __rtruediv__(self, other):
    return self.binary_op(operator.truediv, other, self)

  def __rsub__(self, other):
    return self.binary_op(operator.sub, other, self)

  def __neg__(self):
    return self.map(operator.neg)

  def as_eqs(self):
    return [sympy.Eq(self.sides[i-1], self.sides[i])
            for i in xrange(1, len(self.sides))]
  
  def solve(self, *args, **kwArgs):
    return sympy.solve(self.as_eqs(), *args, **kwArgs)

  def isolate(self, x, lhs=0, rhs=1):
    solution = sympy.solve(self.as_eqs(), x, dict=True)
    if isinstance(solution,list):
      solution, = solution
    return Eqn(x, solution[x])
  
  def __str__(self):
    return ' = '.join(map(str,self.sides))

  def foreach(self, *fs):
    for f in fs:
      self.sides = map(f, self.sides)

  def __getitem__(self,idx):
    return self.sides[idx]

  def __setitem__(self,idx,val):
    self.sides[idx] = val

  @classmethod
  def binary_op(cls, op, a, b):
    if isinstance(b, Eqn) and isinstance(a, Eqn):
      return Eqn(*map(op, a.sides, b.sides))
    if isinstance(a, Eqn):
      return Eqn(*[op(x,b) for x in a.sides])
    assert isinstance(b, Eqn)
    return Eqn(*[op(a,y) for y in b.sides])
  
  def __pow__(self, other):
    return self.binary_op(operator.pow, self, other)
      
  def __mul__(self, other):
    return self.binary_op(operator.mul, self, other)

  def __truediv__(self, other):
    return self.binary_op(operator.truediv, self, other)

  def __add__(self, other):
    return self.binary_op(operator.add, self, other)

  def __sub__(self, other):
    return self.binary_op(operator.sub, self, other)

  def reverse(self):
    return Eqn(*self.sides[::-1])

  def revert(self):
    self.sides = self.sides[::-1]

  def sides_ratio(self, numerator=1, denominator=0):
    return operator.truediv(self.sides[numerator], self.sides[denominator])

  def sides_diff(self, added=1, subtracted=0):
    return self.sides[added]-self.sides[subtracted]

  def subs(self, other):
    return self.map(lambda s: s.subs(other.sides[0], other.sides[1]))

if __name__=='__main__':
  import doctest
  doctest.testmod()  
```

**Comment by oscarbenjamin** [comment]
There is an implementation in the comments here and also in #8023.

I've opened #18053 which makes current `Relational` a proper `Boolean` (not a subclass of `Expr`). What remains is to introduce a new class that implements "symbolic" equations and supports arithmetic operations in the way that users clearly want.

**Comment by oscarbenjamin** [comment]
I didn't realise my comment in #18116 would close this!

**Comment by gutow** [comment]
For my own use I developed an equation class that is just an extension of the classes `Sympy.Expr` and `Sympy.Function`. I think it would be relatively simple to move the extension classes into Symp proper. Presently, I've only completed this for equations using the `=` operator. The extension is less than 150 lines of actual code. Formal unit tests would also have to be written. Presently, tests only exist in the developer and demonstration Jupyter notebooks. Take a look and tell me if this is useful and what people are actually looking for. I would be willing to work on folding this work into SymPy.
[Demonstration in  mybinder](https://mybinder.org/v2/gh/gutow/Algebra_with_Sympy.git/master)
[Github repo](https://github.com/gutow/Algebra_with_Sympy)

**Comment by oscarbenjamin** [comment]
It looks good. One thing I will say is that this should not subclass `Expr` because it does not represent an expression and should not be usable in all contexts where expressions can be used.

This is definitely something that is wanted so if you are interested to work on this then go ahead.

I also wrote the start of an implementation. At the time I found there were problems with the way that `Expr` handled special operations like `__add__` etc so I got distracted fixing those before finishing the `Equation` class itself:
```python
from sympy.core.basic import Basic
from sympy.core.sympify import _sympify
from sympy.simplify import simplify
from sympy.core.relational import Equality

class Equation(Basic):

    def __new__(cls, lhs, rhs):
        lhs = _sympify(lhs)
        rhs = _sympify(rhs)
        return Basic.__new__(cls, lhs, rhs)

    @property
    def lhs(self):
        return self.args[0]

    @property
    def rhs(self):
        return self.args[1]

    def as_Boolean(self):
        return Equality(self.lhs, self.rhs)

    @property
    def reversed(self):
        return Equation(self.rhs, self.lhs)

    def applyfunc(self, func):
        return Equation(func(self.lhs), func(self.rhs))

    def applylhs(self, func):
        return Equation(func(self.lhs), self.rhs)

    def applyrhs(self, func):
        return Equation(self.lhs, func(self.rhs))

    @classmethod
    def _binary_op(cls, a, b, opfunc_ab):
        if isinstance(a, Equation) and not isinstance(b, Equation):
            return Equation(opfunc_ab(a.lhs, b), opfunc_ab(a.rhs, b))
        elif isinstance(b, Equation) and not isinstance(a, Equation):
            return Equation(opfunc_ab(a, b.lhs), opfunc_ab(a, b.rhs))
        elif isinstance(a, Equation) and isinstance(b, Equation):
            return Equation(opfunc_ab(a.lhs, b.lhs), opfunc_ab(a.rhs, b.rhs))
        else:
            raise TypeError('One of a or b should be an equation')

    def __add__(self, other):
        return self._binary_op(self, other, lambda a, b: a + b)

    def __radd__(self, other):
        return self._binary_op(other, self, lambda a, b: a + b)

    def __mul__(self, other):
        return self._binary_op(self, other, lambda a, b: a * b)

    def __rmul__(self, other):
        return self._binary_op(other, self, lambda a, b: a * b)

    def __sub__(self, other):
        return self._binary_op(self, other, lambda a, b: a - b)

    def __rsub__(self, other):
        return self._binary_op(other, self, lambda a, b: a - b)

    def __truediv__(self, other):
        return self._binary_op(self, other, lambda a, b: a / b)

    def __rtruediv__(self, other):
        return self._binary_op(other, self, lambda a, b: a / b)

    def __pow__(self, other):
        return self._binary_op(self, other, lambda a, b: a ** b)

    def __rpow__(self, other):
        return self._binary_op(other, self, lambda a, b: a ** b)

    def __str__(self):
        return '%r = %r' % self.args

    def expand(self, *args, **kwargs):
        return self.func(*(arg.expand(*args, **kwargs) for arg in self.args))

    def simplify(self, *args, **kwargs):
        return simplify(self, *args, **kwargs)

    def _eval_simplify(self, *args, **kwargs):
        return self.func(*(arg.simplify(*args, **kwargs) for arg in self.args))

    def factor(self, *args, **kwargs):
        return factor(self, *args, **kwargs)

    def evalf(self, *args, **kwargs):
        return self.func(*(arg.evalf(*args, **kwargs) for arg in self.args))

    def n(self, *args, **kwargs):
        return self.func(*(arg.n(*args, **kwargs) for arg in self.args))



from sympy import symbols, cos, sin
x, y = symbols('x, y')
eq = Equation(x, y)
print(eq)
eq = Equation(1, 2)
print(eq)
eq = Equation(1, 2) + 3
print(eq)
eq = Equation(1, 2) + Equation(x, y)
print(eq)
eq = Equation(1, 2) - 3
print(eq)
eq = Equation(1, 2) - Equation(x, y)
print(eq)
eq = 1 - Equation(1, 2)
print(eq)
eq = 1 + Equation(x, y)
print(eq)

eq = Equation(cos(x)**2 + sin(x)**2 - 1, 2)
```

**Comment by gutow** [comment]
> One thing I will say is that this should not subclass Expr because it does not represent an expression and should not be usable in all contexts where expressions can be used.

I'm not sure I understand why. Can you provide a specific example of an operation on an expression that you could not apply to both sides of an equation? I can think of instances where it might be silly, but in the most general sense an equation is just another expression with additional operators. It may be that I am thinking of an expression in a different sense than it is actually used in SymPy. I only just started digging into the code to generalize operations I had been doing by hand about 10 days ago. Also this is not a central project for me. I am a Chemistry Professor. I started on this to improve the usability of Jupyter notebooks for simple algebra as part of electronic laboratory notebooks.

**Comment by oscarbenjamin** [comment]
Providing operations for manipulation is not the same as being an expression. It is definitely useful to be able to add something to both sides of the equation and the class I showed above can support that (same as yours):
```julia
In [2]: eq1 = Equation(x, y-1)                                                                                                    

In [3]: eq1                                                                                                                       
Out[3]: x = y - 1

In [4]: eq1 + 1                                                                                                                   
Out[4]: x + 1 = y
```
Allowing arithmetic operations at the Python level does not mean that eq1 *is an expression* though. The distinction is important at the symbolic level when constructing a symbolic object which is not the same as manipulating an object using Python's arithmetic operators.

If the Equation class subclasses `Expr` then it has to be usable *everywhere* `Expr` is e.g. (with your Equation class):
```julia
In [13]: Equation(S(1), S(2)) < 3                                                                                                 
Out[13]: 1=2 < 3

In [14]: Add(Equation(S(1), S(2)), 1)                                                                                             
Out[14]: 1 + 1=2
```
The version I showed would give `TypeError` in each of the above examples but like yours would work with the Python-level operations:
```julia
In [20]: Equation(1, 2) + 1                                                                                                       
Out[20]: 2 = 3
```

**Comment by gutow** [comment]
Thanks. I think I understand your point. Your last example also suggests it would be good to put in some simple sanity checking. It might be nice to flag statements such as `2 > 3` and `2=3` as invalid. I will poke around in the SymPy code base a little more and make another pass. I want to inherit from as much of the class structure as possible, so that I do not have to duplicate existing functionality.

**Comment by oscarbenjamin** [comment]
> It might be nice to flag statements such as `2 > 3` and `2=3` as invalid.

I think that it is best not to check for anything like this. We already have `Equality` etc. which do this as they represent Booleans:
```julia
In [1]: Equality(1, 2)                                                                                                            
Out[1]: False
```
We really need an Equation type that just lets the two sides be. It should be checked that each side sympifies to an expression but otherwise anything goes. We can use something like `eq.as_Boolean` to convert to an `Equality`.

For now I think it's best to have a new Equation class only. I don't think anyone's thought of a good way of handling inequalities and they are less important so best to leave that for the future.

**Comment by gutow** [comment]
The inequalities are definitely more difficult, as they would have to handle assumptions about the symbols properly and throw clearly worded exceptions if there are not enough assumptions to make a valid decision about the outcome of the operation. I had only thought about this a little bit. Unless somebody comes up with a compelling reason equations with inequalities need to be included in SymPy, probably this should simply be left to the heavier duty symbolic math packages (eg. SageMath, Maxima, etc...)

> > It might be nice to flag statements such as 2 > 3 and 2=3 as invalid.

>I think that it is best not to check for anything like this. We already have Equality etc. which do this as they represent Booleans:

This is true, but from a user perspective getting a warning (maybe not an error) if you type something silly is very nice.

**Comment by oscarbenjamin** [comment]
In the broader picture I would like equations to be usable for more than just elementary manipulations (this is not work to be done in one go but looking forwards...)

I think it should be possible to use equations as input and output of functions like solve and also input to subs so e.g.:
```julia
In [30]: eq = [Equation(t, x+y), Equation(z, x-y)]                                                                                

In [31]: eq                                                                                                                       
Out[31]: [t = x + y, z = x - y]

In [32]: eq2 = solve(eq, [x, y])                                                                                                  

In [34]: eq2                                                                                                                      
Out[34]: 
⎡    t   z      t   z⎤
⎢x = ─ + ─, y = ─ - ─⎥
⎣    2   2      2   2⎦

In [35]: f = exp(x+y)*sin(x-y)                                                                                                    

In [37]: f                                                                                                                        
Out[37]: 
 x + y           
ℯ     ⋅sin(x - y)

In [36]: f.subs(eq2)                                                                                       
Out[36]: 
 t       
ℯ ⋅sin(z)
```

**Comment by gutow** [comment]
OK, I see what you are thinking...In that wishlist I would also like linear algebra/matrix/eigenvalue equations...This gets more complicated.  I have not looked at how systems of equations are represented, but I will take a look to make sure not to do something that would obviously break that.

**Comment by mcpl-sympy** [comment]
Seems like we all have similar ideas. My group have been using own Equation class, which is almost identical to gutow's work. The major difference is that my one has `__getattr__` method, which makes thses behavior possible.

```python
>>> eqn = Eqn(f(x), x)
>>> eqn
f(x) = x
>>> eqn.subs(x, 1)
f(1) = 1
>>> eqn.diff(x)
Derivative(f(x), x) = 1
```

**Comment by mcpl-sympy** [comment]
Some additional features that we have implemented:
1. Equation rewriting (the method name is tentative)
```python
>>> eqn = Eqn(x+y, z+1)
>>> eqn
x+y = z + 1
>>> eqn.rewrite(z)
z = 1 - x - y
```

2. Boundary condition
```python
>>> eqn1 = Eqn(f(x), x)
>>> eqn2 = Eqn(x, 0)
>>> BoundaryCondition(eqn1, eqn2)
f(x) = x on x = 0
>>> _.to_eqn()
f(0) = 0
```

Also, we overrided `subs` methods and made some tweaks to SymPy's class signatures so that
```python
>>> (f(x) + 1).subs(eqn)
2
>>> Derivative(eqn, x)
Derivative(f(x), x) = x
```
but this thing may be controversial, I reckon.

**Comment by oscarbenjamin** [comment]
We will definitely want something like the `rewrite` method although it should not be called `rewrite` as that means something else for basic subclasses:
```julia
In [2]: eq = Eq(x, exp(I*t))                                                                                                      

In [3]: eq                                                                                                                        
Out[3]: 
     ⅈ⋅t
x = ℯ   

In [4]: eq.rewrite(cos)                                                                                                           
Out[4]: x = ⅈ⋅sin(t) + cos(t)
```
Rather what is needed is a light-weight solve that only performs invertible operations and only succeeds for a unique result (perhaps having a force parameter). It shouldn't be called `solve` because it should be more limited than `solve` but maybe something like:
```
eq.makelhs(z)
```

Being able to use `Eqn` with `subs` is also good but it should be provided as an iterable of equations (#14741).

We should not modify classes like `Derivative` to handle this but having a `diff` method is fine.

**Comment by gutow** [comment]
I have been working on this and got a lot of this implemented incorporating the ideas from @oscarbenjamin 's [comment above](https://github.com/sympy/sympy/issues/4986#issuecomment-634312860). I may have something for people to look at soon. However, I could use some advice on preferences:
1. I have been using `equ` as the short synonym for `Equation` because I thought it was less likely someone would use that as the name of a variable/function of their own. I noticed in the comments above `eqn` or maybe `Eqn`. What are people's preferences on this?
1. Is there any reason not to extend `solve` for this `Equation` class in exactly the way it is extended for the `Eq` class? All that is done for `Eq` is convert to `Eq.lhs -Eq.rhs` to get the expression into the form expected.
3. How important is the `rewrite` equivalent? Tentatively, I would like to put that off.

**Comment by oscarbenjamin** [comment]
Answering the questions in turn:

1. Users use all possibilities like equ, Eqn etc in their code. We can also count occurrences within the sympy codebase itself:
```
$ git grep '\<equ\>' | wc -l
      12
$ git grep '\<eqn\>' | wc -l
      68
$ git grep '\<eq\>' | wc -l
    2791
$ git grep '\<Eqn\>' | wc -l
      39
```
It should be capitalised as it is a class and ideally it should be short. `Eqn` is probably a good candidate but others might disagree. Note that `Eq` is an short alias for the actual class which is named `Equality`. I'm not sure if we want to do the same with `Eqn/Equation`.

2. I would say leave changes to `solve` for now. Ultimately of course we want to use `Equation` as the (preferred) input type. However I quite like the idea of an API that takes and returns sets of equations e.g.:
```julia
In [7]: sol = {Eqn(t, (x - z)/2), Eqn(y, (x+z)/2)}

In [8]: sol
Out[8]: {t = x/2 - z/2, y = x/2 + z/2}

In [9]: eqns = {Eqn(x, y+t), Eqn(z, y-t)}

In [10]: eqns
Out[10]: {x = t + y, z = -t + y}

In [11]: solve(eqns, {y, t})
Out[12]: [{t = x/2 - z/2, y = x/2 + z/2}]
```
If those same sets of equations are usable as input to `subs` then I think it all links together quite nicely with the idea that a set of equations is the basic object that you can pass around between different APIs:
```julia
In [18]: sol = solve(eqns, {y, t})

In [19]: exp(t - y).subs(sol[0])
Out[19]: 
 -z
ℯ  
```
If the set is also a `FiniteSet` (rather than a Python set) then you can also use subs to substitute *into* the whole set of equations at once:
```julia
In [13]: eqns2 = FiniteSet(*eqns)                                                                                                 

In [14]: eqns2                                                                                                                    
Out[14]: {x = t + y, z = -t + y}

In [15]: eqns2.subs(t, 0)                                                                                                         
Out[15]: {x = y, z = y}
```
I think it would take some time to think about how to make this all work together nicely. The important point though is that right now this is not what `solve` returns but adding a new input type to solve creates an opportunity to change its output type without affecting backward compatibility. The different types returned by `solve` are a complete mess so this opportunity should not be wasted.

For these reasons I suggest holding off on changes to solve until we've had a chance to work these ideas through. We know we want the `Eqn` class for manual manipulations so let's make sure those are working and then we can think about how to combine those usefully in other situations.

3. It's not important to add everything in one go. Actually the most important thing to do now apart from adding the basic class and tests is to clarify the situation in the documentation:

https://docs.sympy.org/latest/tutorial/gotchas.html#equals-signs
https://docs.sympy.org/latest/tutorial/solvers.html#a-note-about-equations

The situation is already confusing and users are already confused so there needs to be somewhere that explains what the different equation classes are for and why we have them. For now also the docs should make it clear that Eqn is a new class which is why most examples seen in other places will use Eq even if Eqn would be better.

Examples throughout the docs should also be updated. This one is from the first page of the tutorial:

https://docs.sympy.org/latest/tutorial/preliminaries.html#installation

**Comment by gutow** [comment]
> `Eqn/Equation`.
I vote for this and will temporarily use it in what I am working on.

**Comment by gutow** [comment]
Based on [Oscar's comment](https://github.com/sympy/sympy/issues/4986#issuecomment-635435428) I will leave solve alone for now. This means I may almost be done. I am struggling with the behavior of integration. Letting the default behavior of differentiation pass through seems fine except for some small inconsistencies in the display of `d` versus `\partial` in some latex expressions (not sure I have time to trace this as I have found the printing system extremely opaque). 

Here is an example of the behavior I have come up with for integration. Is this at all sensible? 
```
In [4]: # Trying to be intelligent about integrating the lhs, but provide choice
   ...: s. 
   ...: q=Eqn(a*c,b/c) 
   ...: q._integ(b)                                                         
Out[4]: Integral(a*c, b)=b**2/(2*c)

In [5]: q._integ(b,eval_lhs=True)                                               
Out[5]: a*b*c=b**2/(2*c)

In [6]: q._integ(c)                                                             
Found one of the integration variables on the lhs. Will integrate
  assuming all other symbols do not depend on integration variables.
  to prevent integration of the lhs, call `._integ` with `eval_lhs=False`.
Out[6]: a*c**2/2=b*log(c)

In [7]: q._integ(c,eval_lhs=False)                                              
Out[7]: Integral(a*c, c)=b*log(c)

In [8]: # integrate passing stuff through _eval_Integral always seems to force e
   ...: valuation.                                                              

In [9]: integrate(q,b)                                                          
WARNING: the lhs will be integrated assuming all symbols are independent of the
    integration variables. If a symbol on the lhs is dependent on an integration
    variable you should use `<eqn_name>._integ(*args, **kwargs, eval_lhs=False)`.
Out[9]: a*b*c=b**2/(2*c)

In [10]: integrate(q,b,eval_lhs=False)                                          
WARNING: the lhs will be integrated assuming all symbols are independent of the
    integration variables. If a symbol on the lhs is dependent on an integration
    variable you should use `<eqn_name>._integ(*args, **kwargs, eval_lhs=False)`.
Out[10]: a*b*c=b**2/(2*c)
```

**Comment by gutow** [comment]
After a little more thought, I think the approach suggested is not going to work. I propose:
1. The default behavior should be to only integrate the rhs and return an unevaluated integral for the lhs. This will be of limited usefulness, but allow getting nice statements such as `f = sin(x)  => Integral(f,x) = -cos(x)`
1. We do not want to do manipulations that will provide results that are inconsistent with the starting equation as in Out[5] below which implies `a = b/(2*c**2)`, rather than ` a=b/c**2` as the original equation defined.
1. The user would have to separately integrate the rhs and lhs with appropriate integration variables, for more sophisticated uses.

In short, I am not sure we want to directly support integration of these equations because of some lack of clarity concerning what the user is trying to achieve.
> 
> Here is an example of the behavior I have come up with for integration. Is this at all sensible?
> 
> ```
> In [4]: # Trying to be intelligent about integrating the lhs, but provide choice
>    ...: s. 
>    ...: q=Eqn(a*c,b/c) 
>    ...: q._integ(b)                                                         
> Out[4]: Integral(a*c, b)=b**2/(2*c)
> 
> In [5]: q._integ(b,eval_lhs=True)                                               
> Out[5]: a*b*c=b**2/(2*c)
> 
> In [6]: q._integ(c)                                                             
> Found one of the integration variables on the lhs. Will integrate
>   assuming all other symbols do not depend on integration variables.
>   to prevent integration of the lhs, call `._integ` with `eval_lhs=False`.
> Out[6]: a*c**2/2=b*log(c)
> 
> In [7]: q._integ(c,eval_lhs=False)                                              
> Out[7]: Integral(a*c, c)=b*log(c)
> 
> In [8]: # integrate passing stuff through _eval_Integral always seems to force e
>    ...: valuation.                                                              
> 
> In [9]: integrate(q,b)                                                          
> WARNING: the lhs will be integrated assuming all symbols are independent of the
>     integration variables. If a symbol on the lhs is dependent on an integration
>     variable you should use `<eqn_name>._integ(*args, **kwargs, eval_lhs=False)`.
> Out[9]: a*b*c=b**2/(2*c)
> 
> In [10]: integrate(q,b,eval_lhs=False)                                          
> WARNING: the lhs will be integrated assuming all symbols are independent of the
>     integration variables. If a symbol on the lhs is dependent on an integration
>     variable you should use `<eqn_name>._integ(*args, **kwargs, eval_lhs=False)`.
> Out[10]: a*b*c=b**2/(2*c)
> ```

**Comment by gutow** [comment]
> After a little more thought, I think the approach suggested is not going to work. I propose:
> 
>     1. The default behavior should be to only integrate the rhs and return an unevaluated integral for the lhs. This will be of limited usefulness, but allow getting nice statements such as `f = sin(x)  => Integral(f,x) = -cos(x)`
> 
>     2. We do not want to do manipulations that will provide results that areyi inconsistent with the starting equation as in Out[5] below which implies `a = b/(2*c**2)`, rather than ` a=b/c**2` as the original equation defined.
> 
>     3. The user would have to separately integrate the rhs and lhs with appropriate integration variables, for more sophisticated uses.
> 
> 
> In short, I am not sure we want to directly support integration of these equations because of some lack of clarity concerning what the user is trying to achieve.

My current implementation of these ideas yields the following:
```
In [3]: q=Eqn(a*c,b/c)                                                          

In [4]: q                                                                       
Out[4]: a*c=b/c

In [5]: integrate(q,b,side='rhs')                                               
Out[5]: b**2/(2*c)

In [6]: integrate(q,b,side='lhs')                                               
Out[6]: a*b*c

In [7]: # Make a pretty statement of integration from an equation 
   ...: Eqn(Integral(q.lhs,b),integrate(q,b,side='rhs'))                        
Out[7]: Integral(a*c, b)=b**2/(2*c)

In [8]: # This is duplicated by the convenience function self.integ 
   ...: q.integ(b)                                                              
Out[8]: Integral(a*c, b)=b**2/(2*c)
```
Unless someone has better ideas, this is what I am going with.

**Comment by gutow** [comment]
I have updated my git repo so people can try my latest version of the Equation class in their own install or using myBinder.org. I will wait a few days for comments before I start thinking about building this into SymPy with formal tests and documentation updates.
[Demonstration in  mybinder](https://mybinder.org/v2/gh/gutow/Algebra_with_Sympy.git/master)
[Github repo](https://github.com/gutow/Algebra_with_Sympy)

**Comment by oscarbenjamin** [comment]
I think you're overthinking the particular example of integrating the equation. We can just do without that for now. Adding something like that is the kind of small change that many people would be interested in contributing once the Eqn class itself exists. If there are details to be ironed out to make it work in a particularly good way then it will be easier to do that later.

When adding the class it is more important to think about how to support manipulations in general. Integration is just one of an enormous number of different operations and manipulations that are performed with expressions that users will want to use. We can't add direct support for all of these so we need to ensure that standard operations are composable with the Eqn class. Broadly the way these operations are done in sympy can be broken down into three categories:

1. Operations like `expr1 + expr2`
2. Function calls like `factor(expr)`.
3. Method calls like `expr.factor()`

Each of these is needed for some operations in sympy so we need to make all three cases work in general. We should start from the position that it isn't possible to make all possible functions and methods work with `Eqn` and that we need to have a general approach for allowing users to perform the above operations to either or both sides of the equation.

1. AFAICT everyone agrees how the operations like + should work. These should apply to both sides of the equation and there is no need to directly provide for say adding 1 to only one side of the equation.

2. Function calls when there are no arguments needed (`factor`, `simplify`, `cancel`, `expand`, etc) are easily provided for by e.g. `eq.applyfunc(factor)`. Where there are additional arguments needed we get the more cumbersome
```python
eq2 = eq1.applylhs(lambda lhs: collect(lhs, x))
```

3. Method calls always need the `lambda`:
```python
eq2 = eq2.applylhs(lambda: lhs.rewrite(cos))
```

We can try to make this easier for users but we need to bear in mind that there is an obvious other way:
```python
eq2 = Eqn(eq2.lhs.rewrite(cos), eq2.rhs)
```
This way of performing a manipulation will work in *all* cases and is easy to understand and not that cumbersome. It's annoying though for operators between two equations in particular e.g. compare:
```python
eq1 = eq2 + eq3
eq1 = Eqn(eq2.lhs + eq2.lhs, eq2.rhs + eq3.rhs)
```
Since I often find myself writing something like that I think it would be nice to be able to do `+` directly. Most manipulations apart from binary operations will not be between two equations meaning that what we have is something more like:
```
eq2 = Eqn(func(eq1.lhs), func(eq1.rhs))
eq2 = Eqn(func(eq1.lhs), eq1.rhs)
eq2 = Eqn(eq1.lhs, func(eq1.rhs))
```

Clearly we will end up supporting some common operations like `diff`/`integrate` in some convenient way but those can always be added later. If there is an integrate method here then it should probably have a signature that is consistent with the interface of `Expr.integrate` because familiarity makes these things easier to understand. Perhaps the `Expr.integrate` method could have an `evaluate=False` option so you can do:
```julia
In [2]: x.integrate(x, evaluate=False)                                                                                            
Out[2]: 
⌠     
⎮ x dx
⌡
```
Then Eqn could have the same so you can do:
```
In [5]: eq = Eqn(f(x), x**2).integrate(x, evaluate=False) 

In [6]: eq
Out[6]: 
⌠           ⌠     
⎮ f(x) dx = ⎮ x dx
⌡           ⌡     

In [8]: eq.applyrhs(lambda rhs: rhs.doit())                                                                                                                 
Out[8]: 
             2
⌠           x 
⎮ f(x) dx = ──
⌡           2 
```

It is possible to make method calling a bit easier using a magic proxy object. I'm not sure if some might object to this idea but it works like this:

1. We have properties on Eqn called do, dolhs, and dorhs.
2. Accessing the properties like `eq.dorhs` returns a special proxy object
3. Calling e.g. `eq.dorhs.doit()` returns `Eqn(eq.lhs, eq.rhs.doit())`.

This makes it possible to call a method on either or both sides of the equation in a way that looks relatively natural:
```
eq.dorhs.factor()  # factor rhs
eq.do.collect(x)   # collect on both sides
eq.do.diff(x)   # etc
eq.do.integrate(x, x)
```
With this method we don't need to add *any* methods to the `Eqn` class. The proxy object simply calls those methods on each side of the equation.

Python being the dynamic language that it is makes it straight-forward to make something like `do` work with a bit of meta-programming but I don't know whether some would object to it being overly magic (plenty of other magic going on in sympy...).

**Comment by oscarbenjamin** [comment]
> I have updated my git repo

The more usual way of getting people to try a new addition to sympy is to open a pull request against this repo and push it up. Right now we are discussing in a long thread that spans 10 years and most contributors are ignoring this or at best skimming our conversation. Once you open a PR more opinions will be forthcoming. It's not necessary for the PR to be work in a final and finished state (it's pretty much guaranteed that changes will be suggested in any case).

**Comment by gutow** [comment]
> The more usual way of getting people to try a new addition to sympy is to open a pull request against this repo and push it up. Right now we are discussing in a long thread that spans 10 years and most contributors are ignoring this or at best skimming our conversation. Once you open a PR more opinions will be forthcoming. It's not necessary for the PR to be work in a final and finished state (it's pretty much guaranteed that changes will be suggested in any case).

I understand your desire to see this really in the code, but as someone who has contributed to a number of large complex projects (the two main ones being [Jmol](http://jmol.sourceforge.net) and [SageMath](https://sagemath.org)) the overhead for putting it into the code, when I can just write extensions of objects is too high until I have something close to what is likely to be accepted. I understand people will need to look at the way things end up in the codebase more seriously, but until I have all the basic concepts mapped out and basically working I am unwilling to devote the time. Until then it is available for use from my git repository (and probably soon via pip). If someone else thinks it is far enough along they are welcome to fold it into their own copy of the codebase.

I believe the myBinder.org demonstration is the easiest way for people to do the initial pass of making sure things are going the correct direction. For example, from your [comment](https://github.com/sympy/sympy/issues/4986#issuecomment-635629566) I need to look more closely at operations like `factor` .

**Comment by gutow** [comment]
> Clearly we will end up supporting some common operations like `diff`/`integrate` in some convenient way but those can always be added later. If there is an integrate method here then it should probably have a signature that is consistent with the interface of `Expr.integrate` because familiarity makes these things easier to understand. Perhaps the `Expr.integrate` method could have an `evaluate=False` option so you can do:

This is already implemented in a way that I believe is consistent with how it is handled by `diff` and `integrate/Integral` (`Integral` is the unevaluated call, as I understand the code) . See [differentiation and integration sections of the github rendering of the Demonstration .ipynb](https://github.com/gutow/Algebra_with_Sympy/blob/master/Demonstration%20of%20algebraic_equation.py.ipynb). I have run into problems with `integrate/Integral` overriding and not passing on keywords. I was only able to get the `side='lhs' or 'rhs'` to pass cleanly. I suspect `evaluate=` is being eaten somewhere else in the code.

**Comment by gutow** [comment]
> This makes it possible to call a method on either or both sides of the equation in a way that looks relatively natural:
> 
> ```
> eq.dorhs.factor()  # factor rhs
> eq.do.collect(x)   # collect on both sides
> eq.do.diff(x)   # etc
> eq.do.integrate(x, x)
> ```

At this point things like
`f.factor()` works to factor both sides
`factor(f.lhs)` returns the factored lhs
`f.lhs.factor()` returns the factored lhs
`f.applylhs(factor)` returns the equation with just the lhs factored.
`factor(f)` presently fails.
This last needs to be fixed, but the others will need a discussion about appropriate syntax, which is what I would like to begin after I get things basically working, and make my first pull request.

**Comment by oscarbenjamin** [comment]
> I understand your desire to see this really in the code

It's more about getting early feedback before spending time on the details. If the end goal is to get this into sympy then it's better to agree things at high-level and then fill out the details later.

**Comment by gutow** [comment]
> > I understand your desire to see this really in the code
> 
> It's more about getting early feedback before spending time on the details. If the end goal is to get this into sympy then it's better to agree things at high-level and then fill out the details later.

Again, I understand this. That is why I started a conversation and am trying to incorporate comments that I get. In particular, your initial idea to back down to the Basic class and your definition of _binary_op() were helpful in making it a little easier to code some of the behaviors. However, from my perspective it is more important to get something working and available in time for me to fold it into my class preparation for next Fall. That means, I need something with reasonably well defined behavior working in the next few weeks. In my experience it is unlikely that I can put in the polishing time necessary to make this become part of SymPy over a period of less than a year. Thus, I first need to write it as an extension. I am still willing and interested in getting this folded into SymPy, but cannot justify working on that until I have something workable for my primary use cases.

If things go well a pull request will come sooner than later. At that point, we can start tweaking to meet the needs of the SymPy project. Writing code is a small component of what I do. My time is primarily devoted to teaching advanced undergraduate chemistry classes and combining quantum chemistry computations with wet lab research into anti-viral compounds.

Regards,
Jonathan

**Comment by mcpl-sympy** [comment]
Implementing `Eqn` class inevitably invokes the necessity of well-defined `evaluate` options all around SymPy. In #18873 and #19377 (closed), I have been trying to integrate all these behaviors. If anyone is interested in this matter, I'd be happy to collaborate.  

My group have been using the custom `Eqn` class we developed to teach chemical material and energy balances in this semester. It worked quite well and proved to be powerfuil. I really do wish to take part in introducing `Eqn` class with similar into SymPy, but this is going to be a huge work.

Again, I'd like to stress that `evaluate` option in method and class signatures, and `doit` method need to be harmonized completely first. Only after then we will be able to do like
```python
>>> eqn = Eqn(f(x), x)

>>> eqn.integrate(x)
Integral(f(x), x) = x**2
>>> eqn.integrate(x, evaluate=False)
Integral(f(x), x) = Integral(x, x)

>>> eqn.subs(x, y)
f(y) = y
>>> eqn.subs(x,y, evaluate=False)
Subs(f(x), x, y) = Subs(x, x, y)
>>> _.doit(term='rhs')
Subs(f(x), x, y) = y
```

**Comment by mcpl-sympy** [comment]
Also, I think `Equation` can be subclass of `Equality` (or vice versa) since they are mathematically relevant. All we need is just make sure that `Equation` is not converted to boolean.

**Comment by oscarbenjamin** [comment]
> Also, I think `Equation` can be subclass of `Equality` (or vice versa) since they are mathematically relevant. All we need is just make sure that `Equation` is not converted to boolean.

We need to ensure that `Equation` is not a subclass of `Boolean`. The purpose of `Equality` is to be used as a `Boolean` and that's why it needs to evaluate in order to simplify things:
```julia
In [2]: p = Piecewise((x, Eq(y, 2)), (z, True))                                                                                   

In [3]: p                                                                                                                         
Out[3]: 
⎧x  for y = 2
⎨            
⎩z  otherwise

In [4]: p.subs(y, 3)                                                                                                              
Out[4]: z
```
That works because
```julia
In [5]: Eq(y, 2).subs(y, 3)                                                                                                       
Out[5]: False
```
This behaviour of `Equality` is documented and is used widely throughout the codebase and so presumably in downstream libraries and user code as well.

I think that the distinction between Eqn as a symbolic representation of a statement and Eq as a Boolean is important internally for sympy although it is possible to have implicit conversions so that users don't need to worry too much about it.

**Comment by gutow** [comment]
I'm inclined to agree with @oscarbenjamin on `Equation` vs `Equality`. The way `Equality` is used in the code does not play well with the concepts underlying the proposed behaviors of `Equation`, nor any of the implementations I have tried.

@mcpl-sympy can you point to your group's definition of the `Eqn` class. I didn't locate it with a quick look at your fork of sympy on Github. I would rather look at the code than have to clone your repository and grep my way through it. Thanks.

**Comment by gutow** [comment]
Before I forget about it again. For the `Eqn` wish list I would add dealing properly with differentials `class sympy.diffgeom.Differential`, which probably means playing nice with all of `class sympy.diffgeom`

**Comment by gutow** [comment]
I have reached the point where I am ready to try adding my Equation class to SymPy. My first thought is to place it in its own directory sympy/sympy/equation/... Is there a better place?

**Comment by oscarbenjamin** [comment]
I'd go with `sympy/core/equation.py`.

**Comment by gutow** [comment]
#19479 is a pull-request providing a working version of my first pass at the `class Equation`.

**Comment by asmeurer** [comment]
Regarding the existing `Eq`, I wonder if we should make it be the equation and have a new class name that is a boolean. There is a lot of existing user code that uses `Eq`, and most of it is using it in an equation sense rather than a boolean equality sense.

**Comment by oscarbenjamin** [comment]
That would be a significant compatibility break for anyone who is using Eq with the intention that it should evaluate to True/False.

Also I quite like the opportunity that introducing a new class brings to be able to redefine the behaviour of some existing functions that currently take Eq. For example we could make it so that `solve` with Eqn inputs has a consistent output type based on list of Eqn. We could also make it so that `subs(listofeqn)` works with a list of Eqn as input. Then the output from solve is consistent and also directly usable with `subs`. If we can eventually deprecate the use of Eq with e.g. `solve` then that provides a pathway to migrate to a more consistent API.

**Comment by oscarbenjamin** [comment]
This is the kind of inconsistency that we could fix:
```python
In [7]: solve([Eq(x, 1), Eq(y, 0)], [x, y])
Out[7]: {x: 1, y: 0}

In [8]: solve([Eq(x**2, 1), Eq(y, 0)], [x, y])
Out[8]: [(-1, 0), (1, 0)]
```

**Comment by gutow** [comment]
> Regarding the existing Eq, I wonder if we should make it be the equation and have a new class name that is a boolean. There is a lot of existing user code that uses Eq, and most of it is using it in an equation sense rather than a boolean equality sense.

>That would be a significant compatibility break for anyone who is using Eq with the intention that it should evaluate to True/False.

I am far from an expert on the SymPy codebase, but based on my working through the code to produce the `Equation` class I agree with @oscarbenjamin. There are places where `Eq` evaluating to True/False is used. I think `Eq` should initially be left and any functionality that duplicates `Equation` can be removed from `Eq`. Once `Eq` can only be used as a boolean, then it would make sense to deprecate it in favor of a more general boolean. At that point maybe `Eq` could become an synonym for `Equation`. I agree `Eq` vs `Eqn` vs `Equation` is currently confusing and others, including me, have used `Eq` for manipulation of equations where the two sides are expressions. But `Eq` does not work well that way because of it's boolean attributes. Thus the need for `Equation`. 

@oscarbenjamin I also agree that getting a consistent return from solve would be very useful. As you've seen if you tried my very preliminary version of `solve(eqn, x)` it returns a list of equations where x is on the lhs.
